### PR TITLE
Skip re-alignment of unchanged traces in `MemAlignOptimizer`

### PR DESCRIPTION
**Pull request by veronicalyu320**
### Related Issues/PRs

N/A

### What changes are proposed in this pull request?

Re-running `MemAlignOptimizer.align()` on already-aligned traces previously double-appended their examples to episodic memory and re-distilled guidelines on top of the prior set. This change makes `align()` idempotent on unchanged traces and correctly handle re-alignment when assessments are edited, added, or removed:

- **Identity re-align (Scenario A)**: re-running `align()` on the exact same traces is now a no-op — no LM call, no index rebuild.
- **Overlap (Scenario B)**: re-running `align()` with a mix of already-aligned traces and new ones only adds the new ones.
- **Content edits (Scenario C)**: when a trace's assessment value/rationale changes, the old episode is dropped and the new one takes its place.
- **Majority-flip (Scenario D)**: when assessment edits change which side wins majority resolution, episodes that left the majority are forgotten and episodes that newly joined are remembered.
- **Assessment removal (Scenario E)**: removing an assessment from a trace and re-aligning forgets the corresponding episode.

Mechanism: `align()` classifies each incoming trace by comparing a `(assessment_id, last_update_time_ms)` fingerprint of its resolved examples against the in-memory representation, then skips unchanged traces, refreshes edited traces in place (drop old episodes, filter guidelines via the existing `source_trace_ids` logic, then add new), and adds new traces. `unalign()` and the in-place refresh path share a new `_apply_unalign_inplace` helper. `trace_to_dspy_example` tags each example with `_trace_id`, `_assessment_id`, and `_last_update_time_ms`, so fingerprints can be derived from existing in-memory state without persisting a parallel structure (no serialization changes).

To retract feedback from a trace, users should continue to use `unalign()` — re-aligning a trace whose assessments are now empty raises (preserves prior behavior).

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

11 new re-alignment tests added in `tests/genai/judges/optimizers/memalign/test_optimizer.py` covering: identity re-align, overlapping batches, in-place content edits, majority-flip on partial / full assessment edits, assessment removal, deserialized-then-realign, mixed-shape calls (unchanged + edited + new in one call), in-batch dedup, unalign→align round-trip, and the preserved-error path for empty-assessment retraction. Combined with the 30 pre-existing memalign tests and 96 sibling optimizer tests, 137 tests pass locally.

### Does this PR require documentation update?

- [x] No.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.
- [ ] Yes. Please link the corresponding PR or explain how you plan to update it.

#### Is this a user-facing change?

- [ ] No.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

`MemAlignOptimizer.align()` is now idempotent when called on traces already aligned, and correctly refreshes memory when traces' assessments are edited, added, or removed.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by forrestmurray-db** [comment]
`_apply_unalign_inplace` drops a guideline only when **all** of its `source_trace_ids` are in the refresh set. When a guideline was distilled from a multi-trace batch and *one* of its source traces is later refreshed, the guideline survives with content derived from the now-stale trace.

### Setup

Two traces aligned together, judge `rag_quality`:

- **T1** — Q: *"What's our refund policy?"* / A: *"We offer 30-day refunds."* → SME label **good**, rationale *"cites the policy directly"*
- **T2** — Q: *"How do I reset my password?"* / A: *"Click forgot password."* → SME label **good**, rationale *"concise and direct"*

### Step 1 — `align(judge, [T1, T2])`

`_distill_new_guidelines` runs the reflection LM over **both** examples in one call and produces a guideline that generalizes across them:

```python
G = Guideline(
    guideline_text="Good answers are direct and concise",
    source_trace_ids=[T1_id, T2_id],   # spans both traces
)
```

State after step 1:

| Episodic memory | Semantic memory |
| --- | --- |
| `T1_ex` (good, "cites policy") | `G` ("direct and concise", sources `[T1, T2]`) |
| `T2_ex` (good, "concise") |  |

### Step 2 — SME edits T1 to **bad** ("missing restocking fees"), then `align(judge_v1, [T1, T2])`

1. Fingerprints: T1 mismatches (`last_update_time_ms` changed) → refresh. T2 matches → skip.
2. `_apply_unalign_inplace({T1_id})` runs the existing filter:

   ```python
   # current: keep G if at least one source survives
   if g.source_trace_ids is None
       or any(tid not in trace_ids_to_remove for tid in g.source_trace_ids)
   ```

   `G.source_trace_ids = [T1, T2]`, `T2 not in {T1}` → **G is retained.**
3. `_add_examples_to_memory([T1_new_ex])` distills from just T1's new example, producing:

   ```python
   G_prime = Guideline(
       guideline_text="Answers should include caveats like restocking fees",
       source_trace_ids=[T1_id],
   )
   ```

State after step 2:

| Episodic memory | Semantic memory |
| --- | --- |
| `T2_ex` (good, "concise") | `G` ("direct and concise", sources `[T1, T2]`) — **stale content** |
| `T1_new_ex` (bad, "missing restocking") | `G'` ("include caveats", sources `[T1]`) |

### Why this matters at inference

Every inference call injects **all** semantic guidelines + retrieved episodic examples. After step 2 the judge sees:

- Guideline: *"Good answers are direct and concise."*
- Retrieved episode (T1_new): *"This direct, concise answer was labeled bad — missing key detail."*
- Guideline: *"Answers should include important caveats."*

The first guideline's text was distilled from a labeling where T1 supported "direct = good". T1 no longer does. The judge has to reconcile the contradiction in-context, and the contradiction is purely a memory-bookkeeping artifact.

### Subsequent `unalign`

Later calling `unalign([T1])`:

- `T1_new_ex` dropped.
- `G'` (sources `[T1]`, all in remove set) → dropped.
- `G` (sources `[T1, T2]`, `T2 not in {T1}`) → **retained**.

Final memory: `[T2_ex]` + `[G with sources=[T1, T2]]`. The guideline cites a `trace_id` that no longer exists in episodic memory.

### Fix (one line in `_apply_unalign_inplace`)

```python
# was: drop G only if ALL sources are in remove-set
if g.source_trace_ids is None
    or any(tid not in trace_ids_to_remove for tid in g.source_trace_ids)

# fix: drop G if ANY source is in remove-set (strict locality)
if g.source_trace_ids is None
    or not any(tid in trace_ids_to_remove for tid in g.source_trace_ids)
```

Tradeoff: the fix loses guidelines that had genuine signal from non-refreshed sources (T2 contributed to "concise" too). That signal is recoverable on a future `align()` round; the alternative — keeping stale-content guidelines in the prompt — actively misleads the judge. Stale > missing isn't a trade I'd take.

### Test 

Add`tests/genai/judges/optimizers/memalign/test_optimizer.py`: 

```python
def test_realign_drops_guidelines_attributed_to_refreshed_trace(
    sample_judge, sample_traces
):
    """Locality: when a trace is refreshed during re-align, any guideline whose
    source_trace_ids includes that trace must be dropped — its content was
    distilled from the trace's prior assessments and is now stale.

    Fails on current code (G is retained because T2 is still a source).
    Passes after _apply_unalign_inplace switches from "drop only if ALL sources
    removed" to "drop if ANY source touched".
    """
    with mock_apis(guidelines=[]):
        optimizer = MemAlignOptimizer()
        judge_v1 = optimizer.align(sample_judge, sample_traces[:2])

        # Simulate the cross-trace guideline that _distill_new_guidelines produces
        # when both traces are aligned in one batch. (Injecting directly so the
        # test is independent of the reflection LM mock's specific output.)
        t0_id = sample_traces[0].info.trace_id
        t1_id = sample_traces[1].info.trace_id
        cross_trace_guideline = Guideline(
            guideline_text="Good answers are direct and concise",
            source_trace_ids=[t0_id, t1_id],
        )
        judge_v1._semantic_memory = [cross_trace_guideline]

        # Edit t0's assessment — flip label and rationale. This bumps
        # last_update_time_ms, so the trace's fingerprint differs from memory
        # and t0 enters the refresh path.
        sample_traces[0].info.assessments = [
            _make_human_assessment(
                assessment_id=None, value="no", rationale="t0 updated"
            ),
        ]

        judge_v2 = optimizer.align(judge_v1, sample_traces[:2])

        # The cross-trace guideline must not survive: t0 was refreshed, so any
        # guideline citing t0 as a source has stale content.
        for g in judge_v2._semantic_memory:
            assert t0_id not in (g.source_trace_ids or []), (
                f"Guideline {g.guideline_text!r} still attributes t0 as a "
                f"source after t0 was refreshed; its content was distilled "
                f"from t0's prior assessments and is now stale."
            )
```

**Comment by veronicalyu320** [comment]
Discussed @forrestmurray-db 's comment [offline](https://databricks.slack.com/archives/C0985DPQRAS/p1777575351091349?thread_ts=1777513706.072859&cid=C0985DPQRAS). The current behavior is the intended behavior.

File: mlflow/genai/judges/optimizers/memalign/optimizer.py [hunk 1]
```diff
@@ -498,6 +498,41 @@ def unalign(self, traces: list[Trace]) -> "MemoryAugmentedJudge":
         )
         return new_judge
 
+    def _in_memory_fingerprint(self, trace_id: str) -> frozenset[tuple[str | None, int | None]]:
+        """Fingerprint of the trace as currently represented in episodic memory.
+        Compared against _examples_fingerprint(...) of freshly-resolved examples
+        from the same trace to decide whether the trace needs refreshing.
+        """
+        return frozenset(
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/memalign/optimizer.py]: this duplicates the `_examples_fingerprint` function, can we have one source of truth for creating a fingerprint?

File: mlflow/genai/judges/optimizers/memalign/optimizer.py [hunk 2]
```diff
@@ -498,6 +498,41 @@ def unalign(self, traces: list[Trace]) -> "MemoryAugmentedJudge":
         )
         return new_judge
 
+    def _in_memory_fingerprint(self, trace_id: str) -> frozenset[tuple[str | None, int | None]]:
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/memalign/optimizer.py:503]: can you remind me how we handle the case where there are multiple assessments on the trace to be aligned on? do we expect users to set up a single aggregate one right now?

File: mlflow/genai/judges/optimizers/memalign/optimizer.py [hunk 3]
```diff
@@ -498,6 +498,41 @@ def unalign(self, traces: list[Trace]) -> "MemoryAugmentedJudge":
...
+            for ex in self._episodic_memory
+            if getattr(ex, "_trace_id", None) == trace_id
+        )
+
+    def _apply_unalign_inplace(self, trace_ids_to_remove: set[str]) -> None:
+        """Drop episodic examples whose trace_id is in the set, prune their entries
+        from _episodic_trace_ids, and filter semantic guidelines via the same
+        source_trace_ids logic. Mutates self. Does not rebuild the episodic
+        search index — the caller decides when to rebuild.
+        """
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/memalign/optimizer.py]:
```suggestion
```

nit: no need to document internal functions, we can use inline comments to explain complex logic

File: mlflow/genai/judges/optimizers/memalign/optimizer.py [hunk 4]
```diff
@@ -675,9 +692,57 @@ def align(self, judge: Judge, traces: list[Trace]) -> MemoryAugmentedJudge:
...
 
-            memory_judge._add_examples_to_memory(new_examples)
+            # Dedupe incoming traces by trace_id (handles align(judge, [t, t])).
+            incoming = {trace.info.trace_id: trace for trace in traces}
+
+            # Build resolved examples once per incoming trace; we reuse them for both
+            # change detection and (if not skipped) memory updates.
+            new_examples_by_trace = {
+                tid: trace_to_dspy_example(trace, judge) for tid, trace in incoming.items()
+            }
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/memalign/optimizer.py]:
```suggestion
            # Build resolved examples once per incoming trace; we reuse them for both
            # change detection and (if not skipped) memory updates.
            examples_by_trace = {
                trace.info.trace_id: trace_to_dspy_example(trace, judge) for trace in traces
            }
```

File: mlflow/genai/judges/optimizers/memalign/optimizer.py [hunk 5]
```diff
@@ -60,6 +60,22 @@ def _get_embedding_batch_size(litellm_model: str) -> int:
     return _DEFAULT_EMBEDDING_BATCH_SIZE
 
 
+def _examples_fingerprint(
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/memalign/optimizer.py]:
```suggestion
def _generate_fingerprint(
```
nit

File: mlflow/genai/judges/optimizers/memalign/optimizer.py [hunk 6]
```diff
@@ -675,9 +692,57 @@ def align(self, judge: Judge, traces: list[Trace]) -> MemoryAugmentedJudge:
...
+                old_fp = memory_judge._in_memory_fingerprint(tid)
+                if not old_fp:
+                    examples_to_align.extend(examples)
+                elif old_fp != new_fp:
+                    trace_ids_to_refresh.add(tid)
+                    examples_to_align.extend(examples)
+                else:
+                    skipped_count += 1
+
+            # Short-circuit when re-aligning on traces already faithfully represented
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/memalign/optimizer.py]: can we simplify this logic to:

if not examples_to_align and skipped_count == 0: no valid feedback
elif not examples_to_align: skipped

File: tests/genai/judges/optimizers/memalign/test_optimizer.py [hunk 7]
```diff
@@ -664,3 +664,320 @@ def test_embedder_batch_size(sample_judge, sample_traces, embedding_model, expec
...
+    return Assessment(
+        name="test_judge",
+        source=AssessmentSource(source_type=AssessmentSourceType.HUMAN, source_id="user1"),
+        feedback=Feedback(value=value),
+        rationale=rationale,
+        assessment_id=assessment_id,
+    )
+
+
+def _make_trace_with_assessments(name: str, assessments: list[Assessment]):
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py]: can we move this to the top and consolidate the way in which we create traces for all tests in this file, not just these ones?

File: tests/genai/judges/optimizers/memalign/test_optimizer.py [hunk 8]
```diff
@@ -664,3 +664,320 @@ def test_embedder_batch_size(sample_judge, sample_traces, embedding_model, expec
...
+    with mlflow.start_span(name=name) as span:
+        span.set_inputs({"inputs": f"input_{name}"})
+        span.set_outputs({"outputs": f"output_{name}"})
+    trace = mlflow.get_trace(mlflow.get_last_active_trace_id())
+    trace.info.assessments = assessments
+    return trace
+
+
+def test_realign_on_same_traces_does_not_duplicate_memory(sample_judge, sample_traces):
+    # Scenario A: re-aligning on the exact same traces should leave memory unchanged.
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py]:
```suggestion
```

File: tests/genai/judges/optimizers/memalign/test_optimizer.py [hunk 9]
```diff
@@ -664,3 +664,320 @@ def test_embedder_batch_size(sample_judge, sample_traces, embedding_model, expec
...
+
+        judge_v2 = optimizer.align(judge_v1, sample_traces[:3])
+
+        assert len(judge_v2._episodic_memory) == v1_episodic_count
+        assert sorted(judge_v2._episodic_trace_ids) == v1_trace_ids
+        assert len(judge_v2._semantic_memory) == v1_semantic_count
+
+
+def test_realign_with_overlapping_traces_only_adds_new(sample_judge, sample_traces):
+    # Scenario B: re-aligning with overlapping + new traces should only add the new ones.
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py]:
```suggestion
```

File: tests/genai/judges/optimizers/memalign/test_optimizer.py (hunk 8 above)
Comment (smoorjani) [reply tests/genai/judges/optimizers/memalign/test_optimizer.py]: same for all tests below - test name is descriptive enough

File: tests/genai/judges/optimizers/memalign/test_optimizer.py [hunk 10]
```diff
@@ -664,3 +664,320 @@ def test_embedder_batch_size(sample_judge, sample_traces, embedding_model, expec
...
+    """Scenario C: re-aligning when the trace's assessment content has changed
+    (same trace_id) should replace the stale example with the updated one.
+    """
+    with mock_apis(guidelines=[]):
+        optimizer = MemAlignOptimizer()
+        judge_v1 = optimizer.align(sample_judge, sample_traces[:3])
+
+        # Mutate trace 0: flip feedback value and rationale (assessment.rationale, the
+        # top-level field that trace_to_dspy_example reads)
+        sample_traces[0].info.assessments = [
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py]: is it possible to use the APIs someone would use to mutate assessment instead of editing inline? the thing that's weird here is that we're explicitly setting assessment id to none which isn't realistic

File: tests/genai/judges/optimizers/memalign/test_optimizer.py [hunk 11]
```diff
@@ -664,3 +664,320 @@ def test_embedder_batch_size(sample_judge, sample_traces, embedding_model, expec
...
+
+def test_realign_updates_majority_resolution(sample_judge):
+    """Scenario D: when assessment edits flip majority membership, examples that were
+    in the prior majority but are no longer should be forgotten, and examples that have
+    newly joined the majority should be remembered.
+    """
+    trace = _make_trace_with_assessments(
+        "majority_trace",
+        [
+            _make_human_assessment("a", "yes", "ra"),
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py]: let's used named arguments (e.g., `_make_human_assessment(assessment_id="a", ...)`) just to make this more readable

File: tests/genai/judges/optimizers/memalign/test_optimizer.py [hunk 12]
```diff
@@ -664,3 +664,320 @@ def test_embedder_batch_size(sample_judge, sample_traces, embedding_model, expec
...
+        v1_semantic_count = len(judge_v1._semantic_memory)
+
+        judge_v2 = optimizer.align(judge_v1, sample_traces[:3])
+
+        assert len(judge_v2._episodic_memory) == v1_episodic_count
+        assert sorted(judge_v2._episodic_trace_ids) == v1_trace_ids
+        assert len(judge_v2._semantic_memory) == v1_semantic_count
+
+
+def test_realign_with_overlapping_traces_only_adds_new(sample_judge, sample_traces):
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py]: this seems redundant with `test_realign_mixed_unchanged_changed_and_new`

File: tests/genai/judges/optimizers/memalign/test_optimizer.py [hunk 13]
```diff
@@ -664,3 +664,320 @@ def test_embedder_batch_size(sample_judge, sample_traces, embedding_model, expec
...
+            ex
+            for ex in judge_v2._episodic_memory
+            if getattr(ex, "_trace_id", None) == sample_traces[1].info.trace_id
+        ]
+        assert len(examples_for_t1) == 1
+        assert examples_for_t1[0].result == "no"
+        assert examples_for_t1[0].rationale == "t1 updated"
+
+
+def test_realign_majority_flips_to_other_side(sample_judge):
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py]: this seems redundant with `test_realign_updates_majority_resolution`

File: tests/genai/judges/optimizers/memalign/test_optimizer.py (hunk 12 above)
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py]: nit: a lot of these tests have the same underlying logic, the setup just changes  a bit. can we see if it's possible to parameterize them?

**Review by smoorjani — CHANGES_REQUESTED** [review]
thanks for updating this! I manually tested and it looks good, but left some comments. would also be good to do some manual tests on your side just to be sure.

File: mlflow/genai/judges/optimizers/memalign/optimizer.py (hunk 2 above)
Comment (veronicalyu320) [reply mlflow/genai/judges/optimizers/memalign/optimizer.py:503]: `trace_to_dspy_example` handles this — it calls `_resolve_assessment_conflicts` (`dspy_utils.py:340`) which groups assessments by feedback value, picks the majority (ties broken by recency), and emits one `dspy.Example` per surviving assessment. Multi-assessment traces work today without users having to aggregate; each surviving assessment becomes its own episode in episodic memory.

File: mlflow/genai/judges/optimizers/memalign/optimizer.py (hunk 6 above)
Comment (veronicalyu320) [reply mlflow/genai/judges/optimizers/memalign/optimizer.py]: There's an edge case that these two logics will behave differently: when user runs align on `[t_already_aligned, t_with_emptied_assessments]`, the existing logic will raise an error, but the simplified logic will silently skip.

Specifically:
- `t_already_aligned` → fingerprint matches → `skipped_count = 1`.
- `t_with_emptied_assessments` → `trace_to_dspy_example` returns `[]` (no human assessments left) → `trace_ids_to_refresh = {tid}`, `examples_to_align` stays `[]`.

End state: `examples_to_align == []`, `skipped_count > 0`, `trace_ids_to_refresh != ∅`. The proposed simplified logic would silently swallow the retraction. Per the pinned design (`test_realign_with_all_empty_assessments_raises`), this case should raise an error — users should call `unalign()` to remove a trace entirely, instead of passing it again with zero assessments.

Went with a flatter form that preserves the current behavior:

```python
if not examples_to_align:
    if trace_ids_to_refresh:
        raise MlflowException(...)   # refresh produced zero examples
    if skipped_count > 0:
        return memory_judge           # all already aligned
    raise MlflowException(...)        # no valid feedback at all
```

File: tests/genai/judges/optimizers/memalign/test_optimizer.py (hunk 12 above)
Comment (veronicalyu320) [reply tests/genai/judges/optimizers/memalign/test_optimizer.py]: Held off for now — the scenarios share shape but each has slightly different setup (different number of assessments, different mutation: update vs delete vs add) and slightly different assertions. A parametrized version would need to express those axes in fixture params and indirect setup helpers, which I think hurts readability more than the duplication does. Happy to revisit if you have a specific shape in mind.

**Comment by veronicalyu320** [comment]
Pushed b712be785 addressing the comments. Highlights:

- merged `_in_memory_fingerprint` into `_generate_fingerprint`, dropped internal docstrings, single dict-comp dedupe, flatter validation (one corner case worth confirming — see https://github.com/mlflow/mlflow/pull/23008#discussion_r3190635987)
- test refactor to use `mlflow.log_feedback` / `update_assessment` / `delete_assessment`; `sample_traces` rewritten to match
- dropped redundant `test_realign_with_overlapping_traces_only_adds_new` and `test_realign_majority_flips_to_other_side`, stripped test docstrings, kwargs in helper calls

Held off on parameterization — see https://github.com/mlflow/mlflow/pull/23008#discussion_r3191248054.

File: mlflow/genai/judges/optimizers/memalign/optimizer.py (hunk 6 above)
Comment (veronicalyu320) [reply mlflow/genai/judges/optimizers/memalign/optimizer.py]: Follow-up: realized the corner I flagged also had a worse sibling. The "use unalign() to retract" rule was being enforced only when *nothing else* was happening — if you mixed an emptied trace with a legitimate new/edited one, the retraction silently went through as a side effect of `_apply_unalign_inplace`. Tightened in dac87d8fb so any incoming trace whose previously-aligned assessments are now empty raises a distinct `Cannot retract feedback by re-aligning with empty assessments` error, regardless of what other traces are in the call. New test `test_realign_with_partial_empty_assessments_raises` covers the mix.

File: mlflow/genai/judges/optimizers/memalign/optimizer.py [hunk 14]
```diff
@@ -675,9 +681,65 @@ def align(self, judge: Judge, traces: list[Trace]) -> MemoryAugmentedJudge:
...
+                    examples_to_align.extend(examples)
+                else:
+                    skipped_count += 1
+
+            if trace_ids_with_empty_assessments:
+                raise MlflowException(
+                    f"Cannot retract feedback by re-aligning with empty assessments. "
+                    f"{len(trace_ids_with_empty_assessments)} trace(s) had their human "
+                    f"assessments removed: {sorted(trace_ids_with_empty_assessments)}. "
+                    f"Use unalign(traces=...) to remove traces from memory.",
```
Comment (forrestmurray-db) [inline mlflow/genai/judges/optimizers/memalign/optimizer.py:746]: This should probably be included in the docstring of align or possibly the optimizer to surface gotchas.

File: mlflow/genai/judges/optimizers/memalign/optimizer.py [hunk 15]
```diff
@@ -498,6 +500,28 @@ def unalign(self, traces: list[Trace]) -> "MemoryAugmentedJudge":
         )
         return new_judge
 
+    def _in_memory_fingerprint(self, trace_id: str) -> frozenset[tuple[str | None, int | None]]:
+        return _generate_fingerprint(
+            ex for ex in self._episodic_memory if getattr(ex, "_trace_id", None) == trace_id
+        )
+
+    def _apply_unalign_inplace(self, trace_ids_to_remove: set[str]) -> None:
+        # Mutates self; does not rebuild the episodic search index (caller's responsibility).
```
Comment (forrestmurray-db) [inline mlflow/genai/judges/optimizers/memalign/optimizer.py:509]: Can we add more explanation here? Initially it wasn't clear that cross-trace distilled guidelines should only be dropped when ALL of their original sources have changed, not ANY.

File: mlflow/genai/judges/optimizers/memalign/optimizer.py (hunk 14 above)
Comment (veronicalyu320) [reply mlflow/genai/judges/optimizers/memalign/optimizer.py:746]: Done!

File: mlflow/genai/judges/optimizers/memalign/optimizer.py (hunk 15 above)
Comment (veronicalyu320) [reply mlflow/genai/judges/optimizers/memalign/optimizer.py:509]: Done!

File: mlflow/genai/judges/optimizers/memalign/optimizer.py [hunk 16]
```diff
@@ -675,9 +704,73 @@ def align(self, judge: Judge, traces: list[Trace]) -> MemoryAugmentedJudge:
...
+                    f"trace(s): {sorted(trace_ids_with_no_feedback)}. Ensure traces contain "
+                    f"human assessments with name '{judge.name}'.",
+                    error_code=INVALID_PARAMETER_VALUE,
+                )
+
+            if not examples_to_align:
+                # Reachable only when every incoming trace was a no-op skip; the
+                # raises above already cover all empty-example paths.
+                _logger.debug("MemAlign alignment skipped: all traces unchanged in memory.")
+                return memory_judge
```
Comment (smoorjani) [inline mlflow/genai/judges/optimizers/memalign/optimizer.py]: should we return the original judge here? we're returning a newly constructed judge which doesn't have episodic memory initialized.

File: tests/genai/judges/optimizers/memalign/test_optimizer.py [hunk 17]
```diff
@@ -664,3 +690,255 @@ def test_embedder_batch_size(sample_judge, sample_traces, embedding_model, expec
...
+        assert len(examples_for_t0) == 1
+        assert examples_for_t0[0].result == "no"
+        assert examples_for_t0[0].rationale == "Updated reason"
+
+
+def test_realign_updates_majority_resolution(sample_judge):
+    trace_id = _start_test_trace("majority_trace")
+    _log_human_feedback(trace_id, value="yes", rationale="ra")
+    _log_human_feedback(trace_id, value="yes", rationale="rb")
+    c = _log_human_feedback(trace_id, value="yes", rationale="rc")
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py]:
```suggestion
    feedback_c = _log_human_feedback(trace_id, value="yes", rationale="rc")
```
style nit: can we use full names? makes it easier to ctrl f in code

File: tests/genai/judges/optimizers/memalign/test_optimizer.py [hunk 18]
```diff
@@ -664,3 +690,255 @@ def test_embedder_batch_size(sample_judge, sample_traces, embedding_model, expec
...
         _, kwargs = mocks["embedder_class"].call_args
         assert kwargs["batch_size"] == expected_batch_size
+
+
+# =============================================================================
+# Re-alignment / deduplication tests
+# =============================================================================
+
+
+def test_realign_on_same_traces_does_not_duplicate_memory(sample_judge, sample_traces):
```
Comment (smoorjani) [inline tests/genai/judges/optimizers/memalign/test_optimizer.py:700]: can we also check that `mocks["lm"]` is never called (from `mock_apis(...) as mocks`)?

Comment (smoorjani) [reply tests/genai/judges/optimizers/memalign/test_optimizer.py:700]: let's also add assertions over # of llm calls in these tests so we can distinguish skipping from re-alignment.

**Review by smoorjani — APPROVED** [review]
left a few comments to address before merging, otherwise LGTM!

File: mlflow/genai/judges/optimizers/memalign/optimizer.py (hunk 16 above)
Comment (veronicalyu320) [reply mlflow/genai/judges/optimizers/memalign/optimizer.py]: Good catch — that was a real bug. `_initialize_dspy_components` deliberately leaves `_retriever = None` (per its inline comment, *"Episodic memory index is not built here — `_add_examples_to_memory()` handles it"*), and the all-skipped short-circuit was returning `memory_judge` without ever building the index, so any `__call__` on the returned judge would have failed inside `retrieve_relevant_examples`. Fixed in fa5d71f5c — now returns the input `judge` directly. Guarded with an `assert isinstance(judge, MemoryAugmentedJudge)` since the skip path can only fire when the input is already a `MemoryAugmentedJudge` (a vanilla `Judge` has no in-memory fingerprints to match against).

File: tests/genai/judges/optimizers/memalign/test_optimizer.py (hunk 17 above)
Comment (veronicalyu320) [reply tests/genai/judges/optimizers/memalign/test_optimizer.py]: Done in fa5d71f5c — also renamed `d` and the matching var in `test_realign_after_assessment_removal` while I was in there.

File: tests/genai/judges/optimizers/memalign/test_optimizer.py (hunk 18 above)
Comment (veronicalyu320) [reply tests/genai/judges/optimizers/memalign/test_optimizer.py:700]: Done in fa5d71f5c. Added reflection-LM `call_count` assertions to the three tests where the distinction is most informative:

- `test_realign_on_same_traces_does_not_duplicate_memory`: snapshots the count after v1 and asserts it's unchanged after the all-skipped v2 (proves the short-circuit really short-circuits).
- `test_realign_replaces_changed_trace_content`: asserts `+1` over v1 (the refreshed example triggers exactly one re-distill).
- `test_align_dedupes_traces_within_a_single_call`: asserts `== 1` total (in-batch dedup collapses the duplicate trace before distillation).

Happy to expand to other tests if useful, but I think these three cover the skip-vs-re-align distinction you're looking for.

**Review by serena-ruan — APPROVED** [review]
